Add ability to override credentials for accessing base backup in S3

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ability to override credentials for accessing base backup in S3 (since tokens may be expired)

Sometimes credentials with which the backup had been done are inactive already, and ClickHouse will not be able to read the metadata file to continue and fail.

Add a setting to allow ignoring credential from base_backup - `use_same_s3_credentials_for_base_backup` (default to true).

And the same for RESTORE.

Cc: @vitlibar

azat committed:
Add ability to override credentials for accessing base backup in S3

Sometimes credentials with which the backup had been done are inactive
already, and ClickHouse will not be able to read the metadata file to
continue and fail.

Add a setting to allow ignoring credential from base_backup -
`use_same_s3_credentials_for_base_backup` (default to true).

And the same for RESTORE.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #53326 from azat/ClickHouse:backups/base-backup-credentials -->
This is an automated comment for commit dc81561de15883927457a078a3b55f7714553e80 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53326/dc81561de15883927457a078a3b55f7714553e80/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

azat commented:
@vitlibar can you take a look?

azat commented:
@vitlibar kind ping

vitlibar commented on src/Backups/BackupSettings.h:
```diff
@@ -44,6 +44,9 @@ struct BackupSettings
     /// Whether native copy is allowed (optimization for cloud storages, that sometimes could have bugs)
     bool allow_s3_native_copy = true;
 
+    /// Whether BACKUP to S3 should use backup_backup field of metadata
```
"backup_backup" seems incorrect

azat replied:
Yep, fixed.

vitlibar commented:
Actually we already have a possibility to override credentials for a base backup:

```
RESTORE ... FROM S3(url, credentials) SETTINGS base_backup=S3(url2, credentials2)
```

vitlibar commented:
> But not for double-incremental:
backup table x to S3(url3) settings base_backup=S3(url2)

As far as I know double-incremental backup is a very rare thing.

vitlibar commented:
I don't like the name `s3_use_credentials_from_base_backup`, it doesn't sound clear. It seems `use_same_s3_credentials_for_base_backup` would be better.

azat commented:
> I don't like the name s3_use_credentials_from_base_backup, it doesn't sound clear.
> It seems use_same_s3_credentials_for_base_backup would be better.

Renamed and rebased.

vitlibar commented on src/Backups/BackupImpl.cpp:
```diff
@@ -166,6 +170,21 @@ void BackupImpl::open(const ContextPtr & context)
+                    writer->substituteCredentials(params.backup_info);
+                    break;
+                case OpenMode::READ:
+                    reader->substituteCredentials(params.backup_info);
```
It is not easy to read this "substituteCredentials()" logic.

I think if we add a function `void BackupInfo::copyS3CredentialsTo(BackupInfo & dest) const`  to `BackupInfo`, and also pass this backup's info to the constructor `BackupImpl` instead of `backup_name_for_logging_` then this code can be rewritten simply as

```
if (base_backup_info)
{
    if (params.use_same_s3_credentials_for_base_backup)
        backup_info.copyS3CredentialsTo(*base_backup_info);

    BackupFactory::CreateParams params;
    params.backup_info = *base_backup_info;
    params.open_mode = OpenMode::READ;
    params.context = context;

    base_backup = BackupFactory::instance().createBackup(params);
    ...
```

azat replied:
Ok, updated.

vitlibar commented on src/Backups/BackupImpl.cpp:
```diff
@@ -166,6 +172,12 @@ void BackupImpl::open(const ContextPtr & context)
+        params.use_same_s3_credentials_for_base_backup = use_same_s3_credentials_for_base_backup;
+
+        if (params.use_same_s3_credentials_for_base_backup)
+            backup_info.copyS3CredentialsTo(params.backup_info);
```
Probably it's better to modify `base_backup_info` here too so

```
BACKUP ... TO S3(url, credentials) SETTINGS base_backup=S3(base_url), use_same_s3_credentials_for_base_backup=true
```

will write `S3(base_url, credentials)` as the base backup to the metadata.

azat replied:
Ok, updated.

azat commented:
>Integration tests (asan, analyzer) [3/6] — fail: 1, passed: 191

- #39700

>Stateless tests (tsan, s3 storage) [5/5] — fail: 1, passed: 1087, skipped: 17

- `00002_log_and_exception_messages_formatting` is flaky in upstream as well

vitlibar commented on src/Backups/BackupInfo.cpp:
```diff
@@ -98,4 +99,23 @@ String BackupInfo::toStringForLogging() const
+    auto & dest_args = dest.args;
+
+    if ((dest_args.size() != 1) && (dest_args.size() != 3))
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Backup S3 requires 1 or 3 arguments: url, [access_key_id, secret_access_key]");
```
It's not a logical error, a logical error means "that cannot happen". In debug version the server even stops on a logical error.

azat replied:
Yes, and here it should never happen, since the same validation exists during parsing backup query, hence LOGICAL_ERROR.

azat replied:
https://github.com/ClickHouse/ClickHouse/blob/55af6e5c3e1bfdee95845279f33a265a9343dca7/src/Backups/registerBackupEngineS3.cpp#L79-L81

azat replied:
>It seems copyS3CredentialsTo is called before creator_fn in registerBackupEngineS3.cpp

Not sure that I got this comment. Can you please elaborate?

vitlibar replied:
I meant `BackupImpl::open()` calls `copyS3CredentialsTo()` before calling `BackupFactory::instance().createBackup()` for the base backup.

azat replied:
Yes, and it should be like that, otherwise it will use incorrect credentials. Not sure that I follow your point.

azat replied:
I guess I got your point about calling `copyS3CredentialsTo` after `createBackup`, please correct me if I'm wrong.

It may throw `LOGICAL_ERROR` in this case for number of arguments check. The latest version of the patch checks number of arguments only for the backup itself (not the `base_backup`), so now, it should not be a problem.

vitlibar replied:
Anyway I think logical error is a too big error for that arguments checking. And it seems better not to put any deep knowledge into this function `copyS3CredentialsTo()` about what and where it's called. `BAD_ARGUMENTS` must be enough for all the checking in this function.

azat replied:
OK

vitlibar commented on src/Backups/BackupInfo.cpp:
```diff
@@ -98,4 +99,23 @@ String BackupInfo::toStringForLogging() const
+    if (dest_args.size() >= 3)
+    {
+        dest_args[1] = args[1];
+        dest_args[2] = args[2];
```
And if `dest_args.size()==1` then we should append two arguments to `dest_args`.

vitlibar replied:
It's necessary to check that `args.size() == 3` as well.

vitlibar commented on src/Backups/BackupInfo.cpp:
```diff
@@ -98,4 +99,23 @@ String BackupInfo::toStringForLogging() const
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Backup S3 requires 1 or 3 arguments: url, [access_key_id, secret_access_key]");
+
+    // 0 - s3_uri
+    if (dest_args.size() >= 3)
```
It can't be `>= 3` because you just checked that it's either `1` or `3`.

azat replied:
This is just a copy-paste of the code that you wrote - https://github.com/ClickHouse/ClickHouse/blob/55af6e5c3e1bfdee95845279f33a265a9343dca7/src/Backups/registerBackupEngineS3.cpp#L79-L84

It can be replaced with `size() == 3`, but I don't see any problem with this notation. If you have some preferences, let me know.

vitlibar commented on src/Backups/BackupSettings.h:
```diff
@@ -44,6 +44,9 @@ struct BackupSettings
     /// Whether native copy is allowed (optimization for cloud storages, that sometimes could have bugs)
     bool allow_s3_native_copy = true;
 
+    /// Whether BACKUP to S3 should use base_backup field of metadata
```
Please correct the comment.

vitlibar commented on tests/queries/0_stateless/02843_backup_use_same_s3_credentials_for_base_backup.sh:
```diff
@@ -0,0 +1,35 @@
+
+function s3_location() { echo "'http://localhost:11111/test/backups/$CLICKHOUSE_DATABASE/use_same_s3_credentials_for_base_backup_base_$*', 'test', 'testtest'"; }
+
+$CLICKHOUSE_CLIENT --format Null -q "BACKUP TABLE data TO S3($(s3_location base))"
```
Why `--format Null` ?

azat replied:
To avoid reporting backup status with id

vitlibar replied:
It would be nice to check the backup status in the test as well (without uuid of course).

azat replied:
Ok, added.

vitlibar commented on tests/queries/0_stateless/02843_backup_use_same_s3_credentials_for_base_backup.sh:
```diff
@@ -0,0 +1,35 @@
+
+function write_invalid_password_to_base_backup()
+{
+    local name=$1 && shift
```
Why is `shift` here?

azat replied:
This is a common practice in bash:

```bash
function foo()
{
    local arg1=$1 && shift
    local arg1=$1 && shift
    ...
}
```

That way you do not rely on the order of args, i.e.

```bash
function foo()
{
    local arg1=$1
    local arg1=$2
    ...
}
```

Here it is not important, but is has zero harm, and later (if) more arguments will be added the diff will contain only one line of change

vitlibar commented on tests/queries/0_stateless/02843_backup_use_same_s3_credentials_for_base_backup.sh:
```diff
@@ -0,0 +1,35 @@
+$CLICKHOUSE_CLIENT --format Null -q "RESTORE TABLE data AS data_1 FROM S3($(s3_location inc_1)) SETTINGS use_same_s3_credentials_for_base_backup=1"
+$CLICKHOUSE_CLIENT --format Null -q "RESTORE TABLE data AS data_2 FROM S3($(s3_location inc_3)) SETTINGS use_same_s3_credentials_for_base_backup=1"
+
+exit 0
```
What's the point of `exit 0` ?

azat replied:
This is a common practice in bash:

```bash
function foo()
{
    local arg1=$1 && shift
    local arg1=$1 && shift
    ...
}
```

That way you do not rely on the order of args, i.e.

```bash
function foo()
{
    local arg1=$1
    local arg1=$2
    ...
}
```

Here it is not important, but is has zero harm, and later (if) more arguments will be added the diff will contain only one line of change

azat replied:
By default bash inherit exit status of the last command, while this is not important for this test, since we have stderr and this is enough to ensure that everything is fine.

Hence explicit `exit 0`

vitlibar commented on tests/queries/0_stateless/02843_backup_use_same_s3_credentials_for_base_backup.sh:
```diff
@@ -0,0 +1,35 @@
+    $CLICKHOUSE_CLIENT --param_content="$content" -q "insert into function s3($(s3_location $name/.backup), 'LineAsString') settings s3_truncate_on_insert=1 values ({content:String})"
+}
+
+function s3_location() { echo "'http://localhost:11111/test/backups/$CLICKHOUSE_DATABASE/use_same_s3_credentials_for_base_backup_base_$*', 'test', 'testtest'"; }
```
Please add a comment for this function `s3_location` with an explanation why it returns what it returns.

azat replied:
Ok, added.

vitlibar commented on src/Backups/BackupInfo.cpp:
```diff
@@ -98,4 +99,29 @@ String BackupInfo::toStringForLogging() const
+    if (backup_engine_name != "S3")
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "use_same_s3_credentials_for_base_backup supported only for S3, got {}", toStringForLogging());
+    if (dest.backup_engine_name != "S3")
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "use_same_s3_credentials_for_base_backup supported only for S3, got {}", dest.toStringForLogging());
```
Again, why is it logical error? If one executes

```
BACKUP ... TO Disk(...) SETTINGS use_same_s3_credentials_for_base_backup = true
```

it's an error, but we should only tell the user that it's an error, not immediately stop the server.

azat replied:
Yep, you are right about checks for `backup_engine_name`. Nice catch. Fixed (and updated the test as well).

vitlibar commented on src/Backups/BackupInfo.cpp:
```diff
@@ -98,4 +98,29 @@ String BackupInfo::toStringForLogging() const
+
+    /// copyS3CredentialsTo() should not be called in this case, since it
+    /// should fail in registerBackupEngineS3(), hence LOGICAL_ERROR
+    if ((args.size() != 1) && (args.size() != 3))
```
Other checks are written in a way that BACKUP/RESTORE will fail if it can't copy credentials.
Perhaps for consistency with other checks it's better to check here

```
if (args.size() != 3)
```
instead of
```
if ((args.size() != 1) && (args.size() != 3))
```

azat replied:
Ok, updated.

azat commented:
>Integration tests (tsan) [5/6] — fail: 1, passed: 447

- `test_merge_tree_azure_blob_storage` - #39700

> Stateless tests (tsan, s3 storage) [5/5] — fail: 1, passed: 1087, skipped: 17

- `00002_log_and_exception_messages_formatting` - failed in upstream as well
